File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/place/PlaceDashboardViewModel.kt
package app.pantopus.android.ui.screens.place

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.place.PlaceIntelligence
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.api.net.displayMessage
import app.pantopus.android.data.homes.HomesRepository
import app.pantopus.android.data.place.PlaceRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.async
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

/**
 * Drives the Place dashboard (C1 verified / C1a claimed). Fetches the
 * living section-envelope payload for a home and exposes the four render
 * states. The home id arrives via [load] (the Home-tab landing renders
 * this screen inline, so there is no nav-arg SavedStateHandle). Mirrors
 * the iOS `PlaceDashboardViewModel`.
 */
@HiltViewModel
class PlaceDashboardViewModel
    @Inject
    constructor(
        private val repo: PlaceRepository,
        private val homesRepository: HomesRepository,
    ) : ViewModel() {
        private val _state = MutableStateFlow<PlaceDashboardUiState>(PlaceDashboardUiState.Loading)
        val state: StateFlow<PlaceDashboardUiState> = _state.asStateFlow()

        private var homeId: String? = null

        /** Bind the screen to a home; idempotent once loaded for that home. */
        fun load(homeId: String) {
            if (this.homeId == homeId && _state.value is PlaceDashboardUiState.Loaded) return
            this.homeId = homeId
            refresh()
        }

        fun refresh() {
            val id = homeId ?: return
            _state.value = PlaceDashboardUiState.Loading
            viewModelScope.launch { fetch(id) }
        }

        private suspend fun fetch(id: String) {
            // The home row rides alongside the intelligence for the one field
            // the dashboard needs from it (move_in_date → the movers card);
            // a failure there never blocks the page.
            val intelligence = viewModelScope.async { repo.intelligence(id) }
            val detail = viewModelScope.async { homesRepository.detail(id) }
            val intelligenceResult = intelligence.await()
            val moveInDate = (detail.await() as? NetworkResult.Success)?.data?.home?.moveInDate
            _state.value =
                when (intelligenceResult) {
                    is NetworkResult.Success -> PlaceDashboardUiState.Loaded(intelligenceResult.data, moveInDate)
                    is NetworkResult.Failure ->
                        PlaceDashboardUiState.Error(intelligenceResult.error.displayMessage("Couldn't load your dashboard."))
                }
        }
    }

sealed interface PlaceDashboardUiState {
    data object Loading : PlaceDashboardUiState

    data class Loaded(
        val intelligence: PlaceIntelligence,
        /** Wedge v2 D5: the movers card shows for ~60 days after this. */
        val moveInDate: String? = null,
    ) : PlaceDashboardUiState

    data class Error(val message: String) : PlaceDashboardUiState
}

File: frontend/apps/ios/Pantopus/Features/Place/PlaceDashboardViewModel.swift
//
//  PlaceDashboardViewModel.swift
//  Pantopus
//
//  Drives the Place dashboard (C1 verified / C1a claimed). Fetches the
//  living section-envelope payload for a home, derives tier + the
//  Today's Pulse hero, and exposes the four render states. Mirrors the
//  `HomeDashboardViewModel` lifecycle and the Android
//  `PlaceDashboardViewModel`.
//

import SwiftUI

@Observable
@MainActor
final class PlaceDashboardViewModel {
    enum State {
        case loading
        case loaded(PlaceIntelligence)
        case error(message: String)
    }

    private(set) var state: State = .loading
    let homeId: String
    /// `Home.move_in_date`, read alongside the intelligence; nil until the
    /// detail fetch lands or when the home has none. Drives `JustMovedCard`.
    private(set) var moveInDate: String?

    private let api: APIClient
    let onOpenDetail: (PlaceDetailGroup) -> Void
    /// Open the full Today's Pulse stream (the hero taps here).
    let onOpenPulse: () -> Void
    /// Switch the dashboard to another of the user's homes.
    let onSelectHome: (String) -> Void
    /// Claim/verify another address (the switcher's "Add a place").
    let onAddPlace: () -> Void
    /// Begin verification with the chosen method + address (pushes B2).
    let onStartVerify: (PlaceVerifyMethod, String) -> Void
    /// W7 — compose a verified-neighbor heads-up (carries the address for
    /// the composer header). Surfaced only for verified (T4) residents.
    let onComposeMessage: (String) -> Void
    /// W7 — open the verified-neighbor inbox.
    let onOpenInbox: () -> Void
    /// Wedge v2 §2 — the privacy mirror ("see what neighbors see").
    let onOpenPrivacyMirror: () -> Void
    /// Movers first (Wedge v2 D5): "Send back the previous resident's mail" opens Mail Day.
    let onOpenMailDay: () -> Void
    let onOpenHubHome: () -> Void

    init(
        homeId: String,
        api: APIClient = .shared,
        onOpenDetail: @escaping (PlaceDetailGroup) -> Void = { _ in },
        onOpenPulse: @escaping () -> Void = {},
        onSelectHome: @escaping (String) -> Void = { _ in },
        onAddPlace: @escaping () -> Void = {},
        onStartVerify: @escaping (PlaceVerifyMethod, String) -> Void = { _, _ in },
        onComposeMessage: @escaping (String) -> Void = { _ in },
        onOpenInbox: @escaping () -> Void = {},
        onOpenPrivacyMirror: @escaping () -> Void = {},
        onOpenMailDay: @escaping () -> Void = {},
        onOpenHubHome: @escaping () -> Void = {}
    ) {
        self.homeId = homeId
        self.api = api
        self.onOpenDetail = onOpenDetail
        self.onOpenPulse = onOpenPulse
        self.onSelectHome = onSelectHome
        self.onAddPlace = onAddPlace
        self.onStartVerify = onStartVerify
        self.onComposeMessage = onComposeMessage
        self.onOpenInbox = onOpenInbox
        self.onOpenPrivacyMirror = onOpenPrivacyMirror
        self.onOpenMailDay = onOpenMailDay
        self.onOpenHubHome = onOpenHubHome
    }

    func load() async {
        if case .loaded = state { return }
        await fetch()
    }

    func refresh() async {
        await fetch()
    }

    private func fetch() async {
        // The move-in date rides on the home detail; it is a nicety, so a
        // failed detail read never blocks the dashboard.
        async let detail: HomeDetailResponse? = try? api.request(HomesEndpoints.detail(homeId: homeId))
        do {
            let intelligence: PlaceIntelligence = try await api.request(
                PlaceEndpoints.intelligence(homeId: homeId)
            )
            // Only a successful detail read may change the date; a failed one
            // keeps the last known value so the card does not blink out.
            if let detail = await detail { moveInDate = detail.home.base.moveInDate }
            state = .loaded(intelligence)
        } catch let error as APIError {
            state = .error(message: error.errorDescription ?? "Couldn't load your place.")
        } catch {
            state = .error(message: "Couldn't load your place.")
        }
    }
}
